fixtures/_fixtures/monkeypatch.py

```diff
@@ -23,21 +23,91 @@
 from fixtures import Fixture
 
 
-def _setattr(obj, name, value):
-    """Handle some corner cases when calling setattr.
+class PartialClassMethod(object):
+    """Emulates the classmethod descriptor, but works on bound methods.
+
+    This wraps a function so that it acts like a classmethod on the calling
+    class. If the wrapped function is already bound to a class then the passed
+    parameters look like foo(boundcls_or_self, callingcls, *args, **kwargs).
 
-    setattr transforms a function into instancemethod, so where appropriate
-    value needs to be wrapped with staticmethod().
     """
+    def __init__(self, f):
+        self.f = f
+
+    def __call__(self, *args, **kwargs):
+        # Called when this is wrapped in another descriptor so __get__ is not
+        # used.
+        return self.f(*args, **kwargs)
+
+    def __get__(self, obj, objtype=None):
+        if objtype is None:
+            objtype = type(obj)
+
+        def newfunc(*args, **kwargs):
+            return self.f(objtype, *args, **kwargs)
+        return newfunc
+
+
+class PartialUnBoundMethod(object):
+    """Emulates an unboundmethod, but works on bound methods.
+
+    This wraps a function so that it acts like an unboundmethod on the calling
+    class. If the wrapped function is already bound to a class then the passed
+    parameters look like foo(boundcls_or_self, callingself, *args, **kwargs).
+
+    """
+    def __init__(self, f):
+        self.f = f
+
+    def __get__(self, obj, objtype=None):
+        def newfunc(*args, **kwargs):
+            return self.f(obj, *args, **kwargs)
+        return newfunc
+
```

Comment:
On Python2 it might, but Python3 removed that as an explicit type in favor of just using FunctionType. I'm not sure of the potential benefit of using that versus returning a partial function here.